presto-hive/src/main/java/com/facebook/presto/hive/HiveDirectoryContext.java

```diff
@@ -13,17 +13,26 @@
  */
 package com.facebook.presto.hive;
 
+import java.util.Map;
+
 import static java.util.Objects.requireNonNull;
 
 public class HiveDirectoryContext
 {
+    public static final String PRESTO_QUERY_ID = "presto_query_id";
+    public static final String PRESTO_QUERY_SOURCE = "presto_query_source";
+    public static final String PRESTO_CLIENT_INFO = "presto_client_info";
+    public static final String PRESTO_USER_NAME = "presto_user_name";
+    public static final String PRESTO_METASTORE_HEADER = "presto_metastore_header";
     private final NestedDirectoryPolicy nestedDirectoryPolicy;
     private final boolean cacheable;
+    private final Map<String, String> additionalProperties;
 
-    public HiveDirectoryContext(NestedDirectoryPolicy nestedDirectoryPolicy, boolean cacheable)
+    public HiveDirectoryContext(NestedDirectoryPolicy nestedDirectoryPolicy, boolean cacheable, Map<String, String> additionalProperties)
     {
         this.nestedDirectoryPolicy = requireNonNull(nestedDirectoryPolicy, "nestedDirectoryPolicy is null");
         this.cacheable = cacheable;
+        this.additionalProperties = requireNonNull(additionalProperties, "additionalProperties is null");
```

Comment:
Do you want to capture an immutable map or live map ? Holding a map means caller can pass in a map and keep mutatating the map. 

If the caller passes in ImmutableMap, calling ImmutableMap.of() has no perf impact.

Reply:
Yeah it makes sense. Even though caller right now is passing immutable map, we can use ImmutableMap.copyOf here to disallow any new caller to mutate.